File: c99/ch01/test_string_search.c
#include <stdio.h>
#include <string.h>

#include "string_search.h"
#include "test.h"

void test_string_search_simple_search(void);
void test_string_search_simple_search_dijkstra_loop(void);

int main(void)
{
    printf("Testing...\n\n");

    printf("string_search_simple_search()\n");
    test_string_search_simple_search();

    printf("string_search_simple_search_dijkstra_loop()\n");
    test_string_search_simple_search_dijkstra_loop();

    return 0;
}

void test_string_search_simple_search(void)
{
    test_assert_equal_size_t(
        string_search_simple_search("test", "test"),  0, "test 01"
    );

    test_assert_equal_size_t(
        string_search_simple_search("test",  "est"),  1, "test 02"
    );
    
    test_assert_equal_size_t(
        string_search_simple_search("test",   "st"),  2, "test 03"
    );
    
    test_assert_equal_size_t(
        string_search_simple_search("test", "t"   ),  0, "test 04"
    );
    
    test_assert_equal_size_t(
        string_search_simple_search("test", "qest"), strlen("test"), "test 05"
    );
}

void test_string_search_simple_search_dijkstra_loop(void)
{
    test_assert_equal_size_t(
        string_search_simple_search_dijkstra_loop("test", "test"),  0, "test 01"
    );

    test_assert_equal_size_t(
        string_search_simple_search_dijkstra_loop("test",  "est"),  1, "test 02"
    );
    
    test_assert_equal_size_t(
        string_search_simple_search_dijkstra_loop("test",   "st"),  2, "test 03"
    );
    
    test_assert_equal_size_t(
        string_search_simple_search_dijkstra_loop("test", "t"   ),  0, "test 04"
    );
    
    test_assert_equal_size_t(
        string_search_simple_search_dijkstra_loop("test", "qest"), strlen("test"), "test 05"
    );
}

File: fortran03/ch01/test_string_search.f90
program test_string_search

  use string_search
  use test

  implicit none

  print *, "Testing..."

  print *, ""
  print *, "string_search_simple_search()"
  call test_string_search_simple_search()

  print *, ""
  print *, "string_search_simple_search_dijkstra_loop()"
  call test_string_search_simple_search_dijkstra_loop()
  
contains

  subroutine test_string_search_simple_search()

    call test_assert_equal(                                              &
      string_search_simple_search("test", "test"), 1, "test 01"          &
    )

    call test_assert_equal(                                              &
      string_search_simple_search("test",  "est"), 2, "test 02"          &
    )
    
    call test_assert_equal(                                              &
      string_search_simple_search("test",   "st"), 3, "test 03"          &
    )
    
    call test_assert_equal(                                              &
      string_search_simple_search("test", "t"   ), 1, "test 04"          &
    )
    
    call test_assert_equal(                                              &
      string_search_simple_search("test", "qest"), 5, "test 05"          &
    )

  end subroutine test_string_search_simple_search
    
  subroutine test_string_search_simple_search_dijkstra_loop()

    call test_assert_equal(                                              &
      string_search_simple_search_dijkstra_loop("test", "test"), 1,      &
        "test 01"                                                        &
    )

    call test_assert_equal(                                              &
      string_search_simple_search_dijkstra_loop("test",  "est"), 2,      &
        "test 02"                                                        &
    )
    
    call test_assert_equal(                                              &
      string_search_simple_search_dijkstra_loop("test",   "st"), 3,      &
        "test 03"                                                        &
    )
    
    call test_assert_equal(                                              &
      string_search_simple_search_dijkstra_loop("test", "t"   ), 1,      &
        "test 04"                                                        &
    )
    
    call test_assert_equal(                                              &
      string_search_simple_search_dijkstra_loop("test", "qest"), 5,      &
        "test 05"                                                        &
    )

  end subroutine test_string_search_simple_search_dijkstra_loop

end program test_string_search
